### PR TITLE
Cleanup error handling in solvers/refinement

### PR DESCRIPTION
**Pull request by xbauch**

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 919e44bbb34668fde6f4cee71d3b65454aa35e18).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84863268

File: src/solvers/refinement/string_constraint_generator_format.cpp [hunk 1]
```diff
@@ -365,7 +366,7 @@ add_axioms_for_format_specifier(
                     << message.eom;
     INVARIANT(
       false, "format specifier must belong to [bBhHsScCdoxXeEfgGaAtT%n]");
-    throw 0;
+    throw unreachable_code_exceptiont("Unknown format specifier.");
```
Comment (peterschrammel) [inline src/solvers/refinement/string_constraint_generator_format.cpp]: That seems an overkill. I'd be fine with `throw 0` in this case.

Comment (romainbrenguier) [reply src/solvers/refinement/string_constraint_generator_format.cpp]: or `UNREACHABLE;` should be fine now

File: src/solvers/refinement/refine_arrays.cpp [hunk 2]
```diff
@@ -100,8 +100,6 @@ void bv_refinementt::arrays_overapproximated()
         false,
         string_refinement_invariantt("error in array over approximation "
           "check"));
-      // Placeholder to tell the compiler we bail
-      throw 0;
```
Comment (tautschnig) [inline src/solvers/refinement/refine_arrays.cpp:104]: I haven't bothered to do a `git blame`, but really the code above these lines should be cleaned up as well: no duplicate text strings, please. The `error()` line should just go away.

File: src/solvers/refinement/string_constraint_generator_main.cpp [hunk 3]
```diff
@@ -554,10 +555,8 @@ std::pair<exprt, string_constraintst> add_axioms_for_char_literal(
   }
   else
   {
-    // convert_char_literal unimplemented
     UNIMPLEMENTED;
-    // For the compiler
-    throw 0;
+    throw unreachable_code_exceptiont("convert_char_literal is unimplemented.");
```
Comment (tautschnig) [inline src/solvers/refinement/string_constraint_generator_main.cpp]: `UNIMPLEMENTED` will render this `throw` unreachable.

File: src/util/exception_utils.cpp [hunk 4]
```diff
@@ -18,3 +18,13 @@ std::string invalid_user_input_exceptiont::what() const noexcept
...
+
+unreachable_code_exceptiont::unreachable_code_exceptiont(std::string message)
+  : message(std::move(message))
+{
+}
+
+std::string unreachable_code_exceptiont::what() const noexcept
+{
+  return message;
+}
```
Comment (tautschnig) [inline src/util/exception_utils.cpp]: No, that's what we've got `UNREACHABLE` for.

**Review by tautschnig — CHANGES_REQUESTED** [review]

**Comment by xbauch** [comment]
Rolled back the introduction of unreachable exception, removed redundant comments.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: cb47dbf2ec1af0d63c341ab8222f4825f39e7735).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85177556

**Review by tautschnig — APPROVED** [review]
Looks good to me modulo squashing the commits and coming up with a combined commit message that explains what has happened.

**Comment by xbauch** [comment]
Squashed into a single commit.

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 60b4986e75b8b7948daeab675d362e4cac328db8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85300087

**Pull request merged**